### PR TITLE
fix build_reference_index items checks

### LINKED ISSUE #1941: Confusing error message
I just inherited gmailr and found that its site hasn't been building for a while.

The problem was an empty item:

https://github.com/r-lib/gmailr/commit/d5d90f30cdf23a44cbaa8fefb22c48370c015f28#diff-f5f791eb2285a738ec729c44088e49a60ddf355eb479fb1fb5df7fcea4fea2e8L75

but the pkgdown error message is a bit misleading. I figure the error with previous pkgdown isn't that interesting, but you can see it [in GHA logs](https://github.com/r-lib/gmailr/runs/4355063527?check_suite_focus=true#step:6:42).

With current dev pkgdown, the error is

```
-- Building function reference -------------------------------------------------
Error : Item 1 in section 8 in reference in '_pkgdown.yml' must be a character.
ℹ You might need to add '' around e.g. - 'N' or - 'off'.
Error: callr subprocess failed: Item 1 in section 8 in reference in '_pkgdown.yml' must be a character.
ℹ You might need to add '' around e.g. - 'N' or - 'off'.
```

This bit was accurate: `section 8 in reference in '_pkgdown.yml'`.

But this was wrong: `Item 1 in ...` (or else maybe I misunderstand what Item 1 refers to).

And this hints feels very specific, but it's also sort of a misdiagnosis:
```
ℹ You might need to add '' around e.g. - 'N' or - 'off'.`
```

### PR DESCRIPTION
**Pull request by maelle**
Fix #1941

### CODE REVIEW

File: R/build-reference-index.R [hunk 1]
```diff
@@ -65,7 +65,25 @@ data_reference_index_rows <- function(section, index, pkg) {
 }
 
 check_all_characters <- function(contents, index, pkg) {
-  any_not_char <- any(purrr::map_lgl(contents, function(x) {typeof(x) != "character"}))
+  null <- purrr::map_lgl(contents, function(x) {typeof(x) == "NULL"})
```
Comment (hadley) [inline R/build-reference-index.R:68]:
```suggestion
  null <- purrr::map_lgl(contents, is.null)
```

File: R/build-reference-index.R [hunk 2]
```diff
@@ -65,7 +65,25 @@ data_reference_index_rows <- function(section, index, pkg) {
...
+          toString(which(null)),
+          index,
+          pkgdown_field(pkg, "reference")
+        ),
+        i = "You might want to delete or fill that line."
+      )
+    )
+  }
+
+  not_char <- purrr::map_lgl(contents, function(x) {typeof(x) != "character"})
```
Comment (hadley) [inline R/build-reference-index.R:85]:
```suggestion
  not_char <- !purrr::map_lgl(contents, is.character)
```

File: R/build-reference-index.R [hunk 3]
```diff
@@ -65,7 +65,25 @@ data_reference_index_rows <- function(section, index, pkg) {
...
+  if (any_null) {
+    abort(
+      c(
+        sprintf(
+          "Item %s in section %s in %s is empty.",
+          toString(which(null)),
+          index,
+          pkgdown_field(pkg, "reference")
+        ),
+        i = "You might want to delete or fill that line."
```
Comment (hadley) [inline R/build-reference-index.R:80]:
```suggestion
        i = "Either delete the empty line or add a function name."
```

**Review by hadley — APPROVED** [review]
Can you add a news bullet too please?

File: R/build-reference-index.R (hunk 1 above)
Comment (maelle) [reply R/build-reference-index.R:68]: :see_no_evil:

**Comment by maelle** [comment]
Oops, didn't see the approval, I might merge it myself tomorrow.

**Comment by hadley** [comment]
Too late 😉

**Pull request merged**